# EDA


- The os module has a perfect method to list files in a directory.
- Pandas json normalize could work here but is not necessary to convert the JSON data to a dataframe.
- You may need a nested for-loop to access each sale!
- We've put a lot of time into creating the structure of this repository, and it's a good example for future projects.  In the file functions_variables.py, there is an example function that you can import and use.  If you have any variables, functions or classes that you want to make, they can be put in the functions_variables.py file and imported into a notebook.  Note that only .py files can be imported into a notebook. If you want to import everything from a .py file, you can use the following:
```python
from functions_variables import *
```
If you just import functions_variables, then each object from the file will need to be prepended with "functions_variables"\
Using this .py file will keep your notebooks very organized and make it easier to reuse code between notebooks.

In [21]:
# (this is not an exhaustive list of libraries)
import pandas as pd
import numpy as np
import os
import json
from pprint import pprint
from functions_variables import encode_tags
from os import listdir
from os.path import join, isfile


## Data Importing

In [29]:
# load one file first to see what type of data you're dealing with and what attributes it has

In [30]:
# loop over all files and put them into a dataframe
import os
import pandas as pd

# Directory containing your JSON files
json_directory = 'C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/data'  # Replace with the actual path to your directory

# Initialize an empty list to store DataFrames from each JSON file
dataframes = []

# Function to load and process each JSON file
def load_json_file(fname):
    """
    This function loads a JSON file (assumed to be ASCII-readable)
    and extracts the 'results' from the 'data' field.
    """
    with open(fname, 'r') as f:
        try:
            data_json = json.load(f)
            # Check if 'data' and 'results' exist in the loaded JSON
            if 'data' in data_json and 'results' in data_json['data']:
                df = pd.json_normalize(data_json['data']['results'])
            else:
                raise KeyError("'data' or 'results' field not found in the JSON file")
        except json.JSONDecodeError as e:
            print(f"Error decoding JSON in file {fname}: {e}")
            return None
        except KeyError as e:
            print(f"Key error in file {fname}: {e}")
            return None
    
    return df

# Loop through all files in the directory
for filename in os.listdir(json_directory):
    if filename.endswith(".json"):  # Check if the file is a JSON file
        file_path = os.path.join(json_directory, filename)  # Full file path
        df = load_json_file(file_path)  # Call load_json_file with fname
        if df is not None:  # If the DataFrame is valid, append to list
            dataframes.append(df)

# Combine all DataFrames into one
if dataframes:
    final_df = pd.concat(dataframes, ignore_index=True)
    print(final_df)
else:
    print("No valid JSON files were processed.")


FileNotFoundError: [Errno 2] No such file or directory: 'C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/data'

In [ ]:
# Save the dataframe to a CSV file
output_csv_path = 'C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/Housing_Data.csv'
final_df.to_csv(output_csv_path, index=False)

print(f"CSV saved to: {output_csv_path}")


CSV saved to: C:/Users/spencer.fargey/LighthouseLabs/tdsf-midterm/Housing_Data.csv


## Data Cleaning and Wrangling

At this point, ensure that you have all sales in a dataframe.
- Take a quick look at your data (i.e. `.info()`, `.describe()`) - what do you see?
- Is each cell one value, or do some cells have lists?
- What are the data types of each column?
- Some sales may not actually include the sale price (target).  These rows should be dropped.
- There are a lot of NA/None values.  Should these be dropped or replaced with something?
    - You can drop rows or use various methods to fills NA's - use your best judgement for each column 
    - i.e. for some columns (like Garage), NA probably just means no Garage, so 0
- Drop columns that aren't needed
    - Don't keep the list price because it will be too close to the sale price. Assume we want to predict the price of houses not yet listed

In [99]:
# load data here, data already concatenated into the csv in previous step
# re import libraries so you can run code from here, and not re-create the original csv file
import pandas as pd
import numpy as np
import os
import json
from pprint import pprint
from functions_variables import encode_tags
from os import listdir
from os.path import join, isfile


df = pd.read_csv("/Users/thomasdoherty/Desktop/tdsf-midterm/Housing_Data.csv") # loading the data from the csv file compiled for thomas' computer

# df = pd.read_csv("/Users/spencer.fargey/Lighthouselabs/tdsf-midterm/Housing_Data.csv") # loading the data from the csv file compiled for Spencers Computer


df.head(5) # make sure it's loaded correctly

last_update_date                                               tags                                          permalink status                    list_date  open_houses                                           branding  list_price   property_id                                             photos  community virtual_tours    listing_id  price_reduced_amount matterport                                 primary_photo.href  source.plan_id                                      source.agents source.spec_id source.type  description.year_built  description.baths_3qtr description.sold_date  description.sold_price  description.baths_full  description.name  description.baths_half  description.lot_sqft  description.sqft  description.baths description.sub_type  description.baths_1qtr  description.garage  description.stories  description.beds description.type lead_attributes.show_contact_an_agent  flags.is_new_construction  flags.is_for_rent  flags.is_subdivision  flags.is_contingent flags.is_price_reduced  flags.is_pending flags.is_foreclosure  flags.is_plan  flags.is_coming_soon flags.is_new_listing products.brand_name                                 other_listings.rdc  location.address.postal_code location.address.state  location.address.coordinate.lon  location.address.coordinate.lat location.address.city location.address.state_code              location.address.line                           location.street_view_url  location.county.fips_code location.county.name  primary_photo  source  products  location.address.coordinate  other_listings community.advertisers community.description.name  location.county
0  2023-09-19T20:52:50Z  ['carport', 'community_outdoor_space', 'cul_de...       9453-Herbert-Pl_Juneau_AK_99801_M90744-30767   sold  2023-06-29T21:16:25.000000Z          NaN  [{'name': 'EXP Realty LLC - Southeast Alaska',...    554950.0  9.074431e+09  [{'tags': [{'label': 'house_view', 'probabilit...        NaN           NaN  2.957242e+09               45000.0      False  https://ap.rdcpix.com/07097d34c98a59ebb7996889...             NaN  [{'office_name': 'EXP Realty LLC - Southeast A...            NaN         mls                  1963.0                     NaN            2023-09-18                     NaN                     2.0               NaN                     NaN               10454.0            1821.0                2.0                  NaN                     NaN                 1.0                  NaN               3.0    single_family                                  True                        NaN                NaN                   NaN                  NaN                  False               NaN                  NaN            NaN                   NaN                False        basic_opt_in  [{'listing_id': '2957241843', 'listing_key': N...                       99801.0                 Alaska                       -134.59372                         58.36395                Juneau                          AK                    9453 Herbert Pl  https://maps.googleapis.com/maps/api/streetvie...                        NaN               Juneau            NaN     NaN       NaN                          NaN             NaN                   NaN                        NaN              NaN
1                   NaN                                                NaN  8477-Thunder-Mountain-Rd_Juneau_AK_99801_M9424...   sold                          NaN          NaN  [{'name': None, 'photo': None, 'type': 'Office'}]         NaN  9.424984e+09                                                NaN        NaN           NaN           NaN                   NaN      False                                                NaN             NaN                                                NaN            NaN         NaN                     NaN                     NaN            2023-08-22                     NaN                     NaN               NaN                     NaN                   NaN               NaN                NaN                  N

In [61]:
# basic exploration of the df

df.shape # 67 possible variables and 8.2k records

(8191, 67)

In [100]:
df.info() # lots of missing values, some completely full columns, some completely empty columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8191 entries, 0 to 8190
Data columns (total 67 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   last_update_date                       8125 non-null   object 
 1   tags                                   7638 non-null   object 
 2   permalink                              8159 non-null   object 
 3   status                                 8159 non-null   object 
 4   list_date                              7752 non-null   object 
 5   open_houses                            0 non-null      float64
 6   branding                               8159 non-null   object 
 7   list_price                             7721 non-null   float64
 8   property_id                            8159 non-null   float64
 9   photos                                 7403 non-null   object 
 10  community                              0 non-null      float64
 11  virt

From just looking at the info above there are **lots of null values**. There doesn't seem any point in getting descriptive statistics until we remove all the completely empty numerical columns. 

Check all columns for completely null

In [101]:
total = df.shape[0] # total number of rows - 8191

# how many missing values are there in each column?
null_counts = df.isnull().sum() # there are many columns with 8191 nulls - the entire column is null

completely_null_cols = null_counts[null_counts == total].index.tolist()

df[completely_null_cols].head(10) # these columns are completely null

open_houses  community  description.name  description.baths_1qtr  flags.is_new_construction  flags.is_for_rent  flags.is_subdivision  flags.is_contingent  flags.is_pending  flags.is_plan  flags.is_coming_soon  primary_photo  source  products  location.address.coordinate  other_listings  location.county
0          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
1          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
2          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
3          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
4          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
5          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
6          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
7          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
8          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN
9          NaN        NaN               NaN                     NaN                        NaN                NaN                   NaN                  NaN               NaN            NaN                   NaN            NaN     NaN       NaN                          NaN             NaN              NaN

In [102]:
df[completely_null_cols].shape # there are 17 columns that are completely null

(8191, 17)

In [103]:
# drop the completely null columns
df = df.drop(columns=completely_null_cols)

In [76]:
df.shape # now is it 50 columns as the 17 completely null are gone

(8191, 50)

We have removed the 17 completely null columns (such as flag.is_subdivision, flag.is_for_rent, location.county...)
Some thoughts:
- The is_subdivision and is_for_rent might have been useful in some predictive way of sale price of a house and they would appear to be yes/no answers (e.g. it is a subdivision / for rent or it is not) but the fact there are no records with a yes (if we make the assumption that NaN means 0, or 'No') means the model wouldn't be able to get any predictive power from that column.

I will look at partially null numerical columns from here and then the categorical columns.

In [104]:
# how many features are numerical and how many are categorical?
numerical = df.select_dtypes(include=[np.number]).columns
categorical = df.select_dtypes(include=['object']).columns

df[numerical].head(15) # there are 37 numerical features

list_price   property_id    listing_id  price_reduced_amount  source.plan_id  description.year_built  description.baths_3qtr  description.sold_price  description.baths_full  description.baths_half  description.lot_sqft  description.sqft  description.baths  description.garage  description.stories  description.beds  location.address.postal_code  location.address.coordinate.lon  location.address.coordinate.lat  location.county.fips_code
0     554950.0  9.074431e+09  2.957242e+09               45000.0             NaN                  1963.0                     NaN                     NaN                     2.0                     NaN               10454.0            1821.0                2.0                 1.0                  NaN               3.0                       99801.0                      -134.593720                        58.363950                        NaN
1          NaN  9.424984e+09           NaN                   NaN             NaN                     NaN                     NaN                     NaN                     NaN                     NaN                   NaN               NaN                NaN                 NaN                  NaN               NaN                       99801.0                              NaN                              NaN                        NaN
2          NaN  9.479069e+09           NaN                   NaN             NaN                     NaN                     NaN                     NaN                     NaN                     NaN                   NaN               NaN                NaN                 NaN                  NaN               NaN                       99801.0                              NaN                              NaN                        NaN
3          NaN  9.879332e+09           NaN                   NaN             NaN                     NaN                     NaN                     NaN                     NaN                     NaN                   NaN               NaN                NaN                 NaN                  NaN               NaN                       99801.0                              NaN                              NaN                        NaN
4          NaN  9.521640e+09           NaN                   NaN             NaN                     NaN                     NaN                     NaN                     NaN                     NaN                   NaN               NaN                NaN                 NaN                  NaN               NaN                       99801.0                              NaN                              NaN                        NaN
5     415000.0  7.412310e+09  2.958925e+09                   NaN             NaN                  1969.0                     NaN                     NaN                     1.0                     NaN                   NaN             950.0                1.0                 NaN                  NaN               2.0                       99801.0                      -134.649067                        58.386651                        NaN
6     835000.0  7.660491e+09  2.958924e+09                   NaN             NaN                  1920.0                     NaN                     NaN                     3.0                     NaN                5875.0            3860.0                3.0                 2.0                  NaN               5.0                       99801.0                      -134.403991                        58.304573                        NaN
7          NaN  8.274121e+09           NaN                   NaN             NaN                  2002.0                     NaN                     NaN                     2.0                     NaN                7476.0            1375.0                2.0                 NaN                  1.0               3.0                       99801.0                      -134.553717                        58.396178                        NaN
8          NaN  9.424984e+

In [105]:
df[numerical].shape # there are 20 numerical features

(8191, 20)

In [94]:
# calculate the proportion of null entries in each column
null_proportion = df.isnull().sum() / total * 100

round(null_proportion.sort_values(ascending=False), 2)

community.description.name               99.94
source.plan_id                           99.94
source.spec_id                           99.94
community.advertisers                    99.94
flags.is_foreclosure                     99.49
description.baths_3qtr                   93.09
virtual_tours                            83.51
description.sub_type                     82.58
description.baths_half                   72.15
flags.is_price_reduced                   69.67
price_reduced_amount                     69.67
description.garage                       45.70
description.stories                      23.57
description.sold_price                   18.01
description.lot_sqft                     14.65
description.baths_full                   10.74
description.year_built                   10.68
description.sqft                         10.60
primary_photo.href                        9.62
photos                                    9.62
description.beds                          8.39
location.coun

Check for proportion of rows with null values when the lat lon is empty

In [106]:
import pandas as pd
import numpy as np

# Step 1: Calculate the total number of rows
total = len(df)

# Step 2: Calculate the proportion of null entries for all rows
null_proportion_all = df.isnull().sum() / total * 100
sorted_null_proportion_all = round(null_proportion_all.sort_values(ascending=False), 2)

# Step 3: Filter rows where both lat and long are null
both_null_condition = df['location.address.coordinate.lat'].isnull() & df['location.address.coordinate.lon'].isnull()
df_both_null = df[both_null_condition]

# Step 4: Calculate the total number of rows where both lat and long are null
total_both_null = len(df_both_null)

# Step 5: Calculate the proportion of null entries where both lat and long are null
null_proportion_both_null = df_both_null.isnull().sum() / total_both_null * 100
sorted_null_proportion_both_null = round(null_proportion_both_null.sort_values(ascending=False), 2)

# Step 6: Combine the two proportions into a DataFrame for side-by-side comparison
comparison_df = pd.DataFrame({
    'Null Proportion (All Rows)': sorted_null_proportion_all,
    'Null Proportion (Lat & Long Null)': sorted_null_proportion_both_null
})

# Adjust display options to show full DataFrame side by side
pd.set_option('display.max_columns', None)  # Show all columns
pd.set_option('display.expand_frame_repr', False)  # Don't break across rows
pd.set_option('display.max_rows', None)  # Show all rows

# Display the DataFrame
print(comparison_df)


                                       Null Proportion (All Rows)  Null Proportion (Lat & Long Null)
branding                                                     0.39                              11.35
community.advertisers                                       99.94                             100.00
community.description.name                                  99.94                             100.00
description.baths                                            2.58                              25.89
description.baths_3qtr                                      93.09                             100.00
description.baths_full                                      10.74                              75.18
description.baths_half                                      72.15                              92.91
description.beds                                             8.39                              75.18
description.garage                                          45.70                          

In [96]:
import pandas as pd
import numpy as np

# Step 1: Filter out rows where both lat and long are null
df_filtered = df[~(df['location.address.coordinate.lat'].isnull() & df['location.address.coordinate.lon'].isnull())]

# Step 2: Fill null list_price values with description.sold_price where description.sold_price is not null
df_filtered['list_price'] = df_filtered['list_price'].fillna(df_filtered['description.sold_price'])

# Step 2: Fill null list_price values with description.sold_price where description.sold_price is not null
df_filtered['list_price'] = df_filtered['description.sold_price'].fillna(df_filtered['list_price'])

# Step 3: Remove rows where list price is null
df_filtered_cleaned = df_filtered[df_filtered['list_price'].notnull()]

# Step 3: Remove rows where list price is null
df_filtered_cleaned = df_filtered[df_filtered['description.sold_price'].notnull()]

# Step 4: Fill null values in description.garage and description.beds with 0
df_filtered_cleaned['description.garage'] = df_filtered_cleaned['description.garage'].fillna(0)
df_filtered_cleaned['description.beds'] = df_filtered_cleaned['description.beds'].fillna(0)

# Step 5: Calculate the total number of rows in the cleaned DataFrame
total_filtered_cleaned = len(df_filtered_cleaned)

# Step 6: Recalculate the proportion of null entries in each column for the cleaned DataFrame
null_proportion_filtered_cleaned = df_filtered_cleaned.isnull().sum() / total_filtered_cleaned * 100

# Step 7: Round and sort the null proportions in descending order
sorted_null_proportion_filtered_cleaned = round(null_proportion_filtered_cleaned.sort_values(ascending=False), 2)

# Display the result
print(sorted_null_proportion_filtered_cleaned)


community.description.name               99.92
source.spec_id                           99.92
community.advertisers                    99.92
source.plan_id                           99.92
flags.is_foreclosure                     99.54
description.baths_3qtr                   94.29
description.sub_type                     82.07
virtual_tours                            81.44
description.baths_half                   69.99
flags.is_price_reduced                   69.42
price_reduced_amount                     69.42
description.stories                      22.44
description.lot_sqft                     12.73
description.baths_full                    8.46
description.year_built                    8.28
description.sqft                          7.88
primary_photo.href                        5.64
photos                                    5.64
tags                                      4.73
location.county.fips_code                 4.67
products.brand_name                       3.95
source.agents

/var/folders/jj/bww8jnqd2lg73nm4k1z5kbmc0000gn/T/ipykernel_31106/2509947596.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['list_price'] = df_filtered['list_price'].fillna(df_filtered['description.sold_price'])
/var/folders/jj/bww8jnqd2lg73nm4k1z5kbmc0000gn/T/ipykernel_31106/2509947596.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['list_price'] = df_filtered['description.sold_price'].fillna(df_filtered['list_price'])
/var/folders/jj/bww8jnqd2lg73nm4k1z5kbmc0000gn/T/ipyker

### Observations: ###
- Almost no entry has a description name, plan_id or spec_id, virtual_tours or 'advertisers' but it's not clear what that would even mean. I want to see what the non-null entries here even look like
- flags.is_foreclosure likely is a yes/no to indicate if this home has been repossessed. Of general housing stock having a very small proportion being foreclosures (0.6% here) seems reasonable so I am confident the nulls here effectively mean 'this home isn't a foreclosure'
- description.baths_3qtr and baths_full probably indicate the number of these structures in the home (same with Garage) so these NaNs can be 0
- What about beds being null (8.3%) does this indicate a studio apartment (technically zero bedrooms)? This is an assumption we will make.
- How is sqft being null (10.6%)? replace these with the mean?
- list_price should be removed once we use any list prices to fill null sale prices - we carry forward with sale price as the best indicator of value

### Actions for numerical column, strikethrough means it is implemented in code below: ###
- list_price:                   Fill sale_price with list_price then drop this. Sale Price is a stronger value metric than list, so use S_P and if it is null, we fill it with L_P
- ~~property_id:                  Drop, serves no analytical purpose, only for a database identifier~~
- ~~listing_id:                   Drop, serves no analytical purpose, only for a database identifier~~
- price_reduced_amount:         Replace NaNs with 0, it is showing how much discount is offered - the % null is identical to flags.is_priced_reduced, so this is the value corresponding to the reduction.
- ~~source.plan_id:               Drop, Probably also a unique identifier with no analytical purpose~~
- description.year_built:       ???
- ~~description.baths_3qtr:       Replace with 0~~
- **description.sold_price:       This column is our prediction variable**
- ~~description.baths_half:       Replace with 0~~
- description.lot_sqft:         ??? Replace with average
- description.sqft:             Replace with average
- ~~description.baths_full:            Replace with 0~~
- ~~description.garage:           Replace with 0~~
- Description.stories:          Replace with 1 - no such thing as a zero story home, assume it means bungalow / one story
- ~~Description.beds:             Replace with 0 - likely indicates a studio apartment~~
- location.address.postal_code: Probably fill in nulls with the postal code which matches the area
- location.address.coord.lon    Drop. Coordinates are not relevant in the combined dataset like they may be in specific cities (e.g. North side nicer than South of a given city) and will distort the model.
- location.address.coord.lat    Drop. See above for longitude.
- ~~location.county.fips_code     Drop~~

In [110]:
# drop ID columns as they have no analytical value
df_filtered_cleaned = df_filtered_cleaned.drop(columns=['property_id', 'listing_id', 'source.plan_id', 'location.county.fips_code']) # dropped ID columns

# replace nulls in description.baths_full, baths_3qtr, garage, beds with 0
df_filtered_cleaned[['description.baths_full', 'description.baths_3qtr', 'description.baths_half', 'description.garage', 'description.beds', 'price_reduced_amount']] = df_filtered_cleaned[['description.baths_full', 'description.baths_3qtr', 'description.baths_half', 'description.garage', 'description.beds', 'price_reduced_amount']].fillna(0) # replaced nulls with 0

df_filtered_cleaned.head(5)

last_update_date                                               tags                                          permalink status                    list_date                                           branding  list_price                                             photos virtual_tours  price_reduced_amount matterport                                 primary_photo.href                                      source.agents source.spec_id source.type  description.year_built  description.baths_3qtr description.sold_date  description.sold_price  description.baths_full  description.baths_half  description.lot_sqft  description.sqft  description.baths description.sub_type  description.garage  description.stories  description.beds description.type lead_attributes.show_contact_an_agent flags.is_price_reduced flags.is_foreclosure flags.is_new_listing products.brand_name                                 other_listings.rdc  location.address.postal_code location.address.state  location.address.coordinate.lon  location.address.coordinate.lat location.address.city location.address.state_code location.address.line                           location.street_view_url location.county.name community.advertisers community.description.name
30  2021-04-13T17:14:34Z  ['central_air', 'central_heat', 'community_out...  501-Moorfield-Ln_Montgomery_AL_36117_M88465-41030   sold  2023-12-19T18:08:22.000000Z  [{'name': 'Forrest Meadows Homes & Land', 'pho...    129900.0                                                NaN           NaN                   0.0      False                                                NaN  [{'office_name': 'Forrest Meadows Homes & Land...            NaN         mls                  1998.0                     0.0            2024-01-16                129900.0                     2.0                     0.0               11761.0            1478.0                2.0                  NaN                 2.0                  1.0               3.0    single_family                                  True                    NaN                  NaN                False          essentials  [{'listing_id': '622475855', 'listing_key': No...                       36117.0                Alabama                       -86.178412                        32.389075            Montgomery                          AL      501 Moorfield Ln  https://maps.googleapis.com/maps/api/streetvie...           Montgomery                   NaN                        NaN
31  2024-01-16T16:07:56Z  ['city_view', 'hardwood_floors', 'two_or_more_...   2326-Winona-Ave_Montgomery_AL_36107_M77279-81021   sold  2023-11-10T14:20:36.000000Z  [{'name': 'Wallace & Moody Realty', 'photo': N...     88500.0  [{'tags': [{'label': 'house_view', 'probabilit...           NaN                3000.0      False  https://ap.rdcpix.com/a5fe1391acf35d43ee6d825c...  [{'office_name': 'Wallace & Moody Realty'}, {'...            NaN         mls                  1945.0                     0.0            2024-01-16                 88500.0                     2.0                     0.0                6534.0            1389.0                2.0                  NaN                 1.0                  2.0               4.0    single_family                                  True                  False                  NaN                False          essentials  [{'listing_id': '2961522977', 'listing_key': N...                       36107.0                Alabama                       -86.273286                        32.382748            Montgomery                          AL       2326 Winona Ave  https://maps.googleapis.com/maps/api/streetvie...           Montgomery                   NaN                        NaN
32  2021-04-13T17:14:32Z  ['carport', 'central_air', 'central_heat', 'di...  115-Lookout-Ridge-Rd_Montgomery_AL_36109_M7320...   sold  2023-12-19T18:08:22.000000Z  [{'name': 'FlatFee.com', 'photo': None, 'type'...    145000.0                                                NaN           NaN       

In [111]:
# verify no nulls in the replaced columns in df_filtered_cleaned
df_filtered_cleaned[['description.baths_full', 'description.baths_3qtr', 'description.baths_half', 'description.garage', 'description.beds', 'price_reduced_amount']].isnull().sum()

description.baths_full    0
description.baths_3qtr    0
description.baths_half    0
description.garage        0
description.beds          0
price_reduced_amount      0
dtype: int64

In [58]:
df.head(5)

last_update_date                                               tags                                          permalink status                    list_date                                           branding  list_price                                             photos virtual_tours  price_reduced_amount matterport                                 primary_photo.href                                      source.agents source.spec_id source.type  description.year_built  description.baths_3qtr description.sold_date  description.sold_price  description.baths_full  description.baths_half  description.lot_sqft  description.sqft  description.baths description.sub_type  description.garage  description.stories  description.beds description.type lead_attributes.show_contact_an_agent flags.is_price_reduced flags.is_foreclosure flags.is_new_listing products.brand_name                                 other_listings.rdc  location.address.postal_code location.address.state  location.address.coordinate.lon  location.address.coordinate.lat location.address.city location.address.state_code              location.address.line                           location.street_view_url  location.county.fips_code location.county.name community.advertisers community.description.name
0  2023-09-19T20:52:50Z  ['carport', 'community_outdoor_space', 'cul_de...       9453-Herbert-Pl_Juneau_AK_99801_M90744-30767   sold  2023-06-29T21:16:25.000000Z  [{'name': 'EXP Realty LLC - Southeast Alaska',...    554950.0  [{'tags': [{'label': 'house_view', 'probabilit...           NaN               45000.0      False  https://ap.rdcpix.com/07097d34c98a59ebb7996889...  [{'office_name': 'EXP Realty LLC - Southeast A...            NaN         mls                  1963.0                     NaN            2023-09-18                     NaN                     2.0                     NaN               10454.0            1821.0                2.0                  NaN                 1.0                  NaN               3.0    single_family                                  True                  False                  NaN                False        basic_opt_in  [{'listing_id': '2957241843', 'listing_key': N...                       99801.0                 Alaska                       -134.59372                         58.36395                Juneau                          AK                    9453 Herbert Pl  https://maps.googleapis.com/maps/api/streetvie...                        NaN               Juneau                   NaN                        NaN
1                   NaN                                                NaN  8477-Thunder-Mountain-Rd_Juneau_AK_99801_M9424...   sold                          NaN  [{'name': None, 'photo': None, 'type': 'Office'}]         NaN                                                NaN           NaN                   NaN      False                                                NaN                                                NaN            NaN         NaN                     NaN                     NaN            2023-08-22                     NaN                     NaN                     NaN                   NaN               NaN                NaN                  NaN                 NaN                  NaN               NaN              NaN                                  True                    NaN                  NaN                  NaN                 NaN  [{'listing_id': '2958935271', 'listing_key': N...                       99801.0                 Alaska                              NaN                              NaN                Juneau                          AK           8477 Thunder Mountain Rd  https://maps.googleapis.com/maps/api/streetvie...                        NaN               Juneau                   NaN                        NaN
2                   NaN                                                NaN      4515-Glacier-Hwy_Juneau_AK_99801_M94790-68516   sold                          NaN  [{'name': None, 

In [78]:
df[categorical].shape # there are 30 categorical features

(8191, 30)

In [79]:
df[categorical].head(5)

last_update_date                                               tags  \
0  2023-09-19T20:52:50Z  ['carport', 'community_outdoor_space', 'cul_de...   
1                   NaN                                                NaN   
2                   NaN                                                NaN   
3                   NaN                                                NaN   
4                   NaN                                                NaN   

                                           permalink status  \
0       9453-Herbert-Pl_Juneau_AK_99801_M90744-30767   sold   
1  8477-Thunder-Mountain-Rd_Juneau_AK_99801_M9424...   sold   
2      4515-Glacier-Hwy_Juneau_AK_99801_M94790-68516   sold   
3  17850-Point-Stephens-Rd_Juneau_AK_99801_M98793...   sold   
4  9951-Stephen-Richards-Memorial-Dr_Juneau_AK_99...   sold   

                     list_date  \
0  2023-06-29T21:16:25.000000Z   
1                          NaN   
2                          NaN   
3                          NaN   
4                          NaN   

                                            branding  \
0  [{'name': 'EXP Realty LLC - Southeast Alaska',...   
1  [{'name': None, 'photo': None, 'type': 'Office'}]   
2  [{'name': None, 'photo': None, 'type': 'Office'}]   
3  [{'name': None, 'photo': None, 'type': 'Office'}]   
4  [{'name': None, 'photo': None, 'type': 'Office'}]   

                                              photos virtual_tours matterport  \
0  [{'tags': [{'label': 'house_view', 'probabilit...           NaN      False   
1                                                NaN           NaN      False   
2                                                NaN           NaN      False   
3                                                NaN           NaN      False   
4                                                NaN           NaN      False   

                                  primary_photo.href  ... products.brand_name  \
0  https://ap.rdcpix.com/07097d34c98a59ebb7996889...  ...        basic_opt_in   
1                                                NaN  ...                 NaN   
2                                                NaN  ...                 NaN   
3                                                NaN  ...                 NaN   
4                                                NaN  ...                 NaN   

                                  other_listings.rdc location.address.state  \
0  [{'listing_id': '2957241843', 'listing_key': N...                 Alaska   
1  [{'listing_id': '2958935271', 'listing_key': N...                 Alaska   
2  [{'listing_id': '2958935192', 'listing_key': N...                 Alaska   
3  [{'listing_id': '2958925235', 'listing_key': N...                 Alaska   
4  [{'listing_id': '2958924367', 'listing_key': N...                 Alaska   

  location.address.city location.address.state_code  \
0                Juneau                          AK   
1                Juneau                          AK   
2                Juneau                          AK   
3                Juneau                          AK   
4                Juneau                          AK   

               location.address.line  \
0                    9453 Herbert Pl   
1           8477 Thunder Mountain Rd   
2                   4515 Glacier Hwy   
3            17850 Point Stephens Rd   
4  9951 Stephen Richards Memorial Dr   

                            location.street_view_url location.county.name  \
0  https://maps.googleapis.com/maps/api/streetvie...               Juneau   
1  https://maps.googleapis.com/maps/api/streetvie...               Juneau   
2  https://maps.googleapis.com/maps/api/streetvie...               Juneau   
3  https://maps.googleapis.com/maps/api/streetvie...               Juneau   
4  https://maps.googleapis.com/maps/api/streetvie...               Juneau   

  community.advertisers community.description.name  
0                   NaN                        NaN  
1                   NaN    

### Actions for categorical columns ###

- last_update_date:         Drop
- tags:                     More info required
- permalink:                Drop
- status                    almost all complete / what are nulls?
- list_date                 Take the mean if there is a null? check % of nulls Days since listing could be interesting to find?
- branding                  Drop
- photos                    Drop - its a representation of other features in the home
- virtual_tours             ??? more info required
- matterport                ???
- primary_photo.href        Drop
- products.brand_name       ???
- other_listings.rdc        ??? inclined to drop
- location.address.state    Inclined to drop - different states are not able to be numerically represented
- location.address.city     Could be Feature Engineered - find city population and add this as a column? the city is probably important for house price, how to we numerically represent?
- location.address.line     Drop
- location.street_view_url  Drop
- location.county.name      Fill in if city is known
- community.advertisers     Drop
- community.description.name Drop or more into required



### Dealing with Tags

Consider the fact that with tags, there are a lot of categorical variables.
- How many columns would we have if we OHE tags, city and state?
- Perhaps we can get rid of tags that have a low frequency.

In [ ]:
# OHE categorical variables/ tags here
# tags will have to be done manually

### Dealing with Cities

- Sales will vary drastically between cities and states.  Is there a way to keep information about which city it is without OHE?
- Could we label encode or ordinal encode?  Yes, but this may have undesirable effects, giving nominal data ordinal values.
- What we can do is use our training data to encode the mean sale price by city as a feature (a.k.a. Target Encoding)
    - We can do this as long as we ONLY use the training data - we're using the available data to give us a 'starting guess' of the price for each city, without needing to encode city explicitly
- If you replace cities or states with numerical values (like the mean price), make sure that the data is split so that we don't leak data into the training selection. This is a great time to train test split. Compute on the training data, and join these values to the test data
- Note that you *may* have cities in the test set that are not in the training set. You don't want these to be NA, so maybe you can fill them with the overall mean

In [ ]:
# perform train test split here
# do something with state and city

## Extra Data - STRETCH

> This doesn't need to be part of your Minimum Viable Product (MVP). We recommend you write a functional, basic pipeline first, then circle back and join new data if you have time

> If you do this, try to write your downstream steps in a way it will still work on a dataframe with different features!

- You're not limited to just using the data provided to you. Think/ do some research about other features that might be useful to predict housing prices. 
- Can you import and join this data? Make sure you do any necessary preprocessing and make sure it is joined correctly.
- Example suggestion: could mortgage interest rates in the year of the listing affect the price? 

In [ ]:
# import, join and preprocess new data here

## EDA/ Visualization

Remember all of the EDA that you've been learning about?  Now is a perfect time for it!
- Look at distributions of numerical variables to see the shape of the data and detect outliers.    
    - Consider transforming very skewed variables
- Scatterplots of a numerical variable and the target go a long way to show correlations.
- A heatmap will help detect highly correlated features, and we don't want these.
    - You may have too many features to do this, in which case you can simply compute the most correlated feature-pairs and list them
- Is there any overlap in any of the features? (redundant information, like number of this or that room...)

In [ ]:
# perform EDA here

## Scaling and Finishing Up

Now is a great time to scale the data and save it once it's preprocessed.
- You can save it in your data folder, but you may want to make a new `processed/` subfolder to keep it organized